tokenizer_ts/src/bytePairEncode.ts

```diff
@@ -1,15 +1,66 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
-/**
- * Convert a Uint8Array to a string
- * @param uint8Array
- * @returns string
- */
-export function uint8ArrayToString(uint8Array: Uint8Array): string {
-  return Array.from(uint8Array)
-    .map(num => num.toString())
-    .join("_");
+const enum Constant {
+  // we have 48 bits per level, we can safely bitwise encode 32 bits at a time,
+  // so this works in two passes
+  BytesPerLevel = 6,
```

Comment:
Not sure how you got 48 bits per level, could you explain that?

Reply:
That's the number of full bytes that fit into the 52 bits is Number.MAX_SAFE_INTEGER before it starts to lose precision

Reply:
Does this mean that there is a max 8 levels to the map?

Reply:
There's infinite levels, but each level can only encode 6 bytes. The longest byte pair prefix in the 100k token model is 128 bytes which would go 22 levels deep, but most lookups are much shallower than that.

Reply:
Right- we're just limited in BytesPerLevel based on the number of bits in Number.MAX_SAFE_INTEGER (52), but the Uint8Array can technically be of infinite length?

Reply:
It can, yea. Though if we get a lookup and there's no extra map level where we're looking we exit early and return undefined